merlin/models/torch/blocks/experts.py

```diff
@@ -0,0 +1,244 @@
+    learning (mtl) model for personalized recommendations."
+    Fourteenth ACM Conference on Recommender Systems. 2020.
+"""
+
+
+class MMOEBlock(Block):
+    """
+    Multi-gate Mixture-of-Experts (MMoE) Block introduced in [1].
+
+    References
+    ----------
+    [1] Ma, Jiaqi, et al. "Modeling task relationships in multi-task learning with
+    multi-gate mixture-of-experts." Proceedings of the 24th ACM SIGKDD international
+    conference on knowledge discovery & data mining. 2018.
+
+    Parameters
+    ----------
+    expert : nn.Module
+        The base expert model to be used.
+    num_experts : int
+        The number of experts to be used.
+    outputs : Optional[ParallelBlock]
+        The output block. If it is an instance of ParallelBlock,
+        repeat it for each expert, otherwise use a single ExpertGateBlock.
+    """
+
+    def __init__(
+        self, expert: nn.Module, num_experts: int, outputs: Optional[ParallelBlock] = None
+    ):
+        experts = repeat_parallel(expert, num_experts, agg=Stack(dim=1))
+        super().__init__(ShortcutBlock(experts, output_name="experts"))
+        if isinstance(outputs, ParallelBlock):
+            self.append(repeat_parallel_like(ExpertGateBlock(num_experts), outputs))
+        else:
+            self.append(ExpertGateBlock(num_experts))
+
+
+@docstring_parameter(ple_reference=_PLE_REFERENCE)
+class PLEBlock(Block):
+    """
+    Progressive Layered Extraction (PLE) Block  proposed in [1].
+
+    {ple_reference}
+
+    Parameters
+    ----------
+    expert : nn.Module
+        The base expert model to be used.
+    num_shared_experts : int
+        The number of shared experts.
+    num_task_experts : int
+        The number of task-specific experts.
+    depth : int
+        The depth of the network.
+    outputs : ParallelBlock
+        The output block.
+    """
+
+    def __init__(
+        self,
+        expert: nn.Module,
+        *,
+        num_shared_experts: int,
+        num_task_experts: int,
+        depth: int,
+        outputs: ParallelBlock,
+    ):
+        cgc_kwargs = {
+            "expert": expert,
+            "num_shared_experts": num_shared_experts,
+            "num_task_experts": num_task_experts,
+            "outputs": outputs,
+        }
+        super().__init__(*CGCBlock(shared_gate=True, **cgc_kwargs).repeat(depth - 1))
+        self.append(CGCBlock(**cgc_kwargs))
+
+
+class CGCBlock(Block):
+    """
+    Implements the Customized Gate Control (CGC) proposed in [1].
+
+    {ple_reference}
+
+    Parameters
+    ----------
+    expert : nn.Module
+        The base expert model to be used.
+    num_shared_experts : int
+        The number of shared experts.
+    num_task_experts : int
+        The number of task-specific experts.
+    outputs : ParallelBlock
+        The output block.
+    shared_gate : bool, optional
+        If true, use a shared gate for all tasks. Defaults to False.
```

Comment:
I think this argument name and docstrings is misleading. When `num_shared_experts>0` all task gates will use the shared experts, right?
This `shared_gate` argument seems to be responsible to make it available the output of the shared expert (`shortcut` and `experts` keys) from one CGCBlock to the next one in PLE architecture? If that is the case, we could try and clarify that.